File: src/leetcode/second/Code_160_IntersectionOfLinkedList.java
package leetcode.second;

import tool.LinkedListTool;
import tool.ListNode;

/**
 * 编写一个程序，找到两个单链表相交的起始节点。
 *
 * @since 2019-12-27 Friday
 */
public class Code_160_IntersectionOfLinkedList {
    static ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) {
            return null;
        }
        ListNode l1 = headA, l2 = headB;
        while (l1 != l2) {
            l1 = l1 == null ? headB : l1.next;
            l2 = l2 == null ? headA : l2.next;
        }
        return l1;
    }

    public static void main(String[] args) {
        ListNode l1 = LinkedListTool.generateList(new int[]{1, 2, 3});
        ListNode l2 = LinkedListTool.generateList(new int[]{1, 2, 3});
        System.out.println(getIntersectionNode(l1, l2));
    }
}

File: src/cpp-leetcode/first/Code_160_IntersectionOfLinkedlist.cpp
#include <iostream>
#include <vector>

#include "../List.h"

/**
 * intersection of linked-list
 *
 * @since 2020-9-16 Wednesday 15:44
 */
class Solution {
   public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (headA == nullptr || headB == nullptr) {
            return nullptr;
        }
        ListNode *l1 = headA, *l2 = headB;
        while (l1 != l2) {
            l1 = l1 == nullptr ? headB : l1->next;
            l2 = l2 == nullptr ? headA : l2->next;
        }
        return l1;
    }
};

int main(int argc, char const *argv[]) {
    Solution s;
    std::vector<int> v = {1, 2, 3, 4};
    ListNode *headA = List::linkedList(v);
    ListNode *headB = new ListNode(9);
    headB->next = headA->next->next;
    ListNode *res = s.getIntersectionNode(headA, headB);
    if (res == NULL) {
        std::cout << "no intersection\n";
        exit(-1);
    }
    std::cout << res->val << std::endl;
}
